lambda-buffers-compiler/test/Test/TypeClassCheck.hs

```diff
@@ -70,3 +101,219 @@ nocycles =
   , mkclass "Monad" ["Applicative"]
   , mkclass "Traversable" ["Foldable", "Functor"]
   ]
+
+---- Solver tests
+
+solveTests :: TestTree
+solveTests =
+  testGroup
+    "Solver tests"
+    [ testCase "1: C [Maybe Int] (completeRules)" $
+        solveTest1 @?= []
+    , testCase "2: D [Maybe Int] (partialRules)" $
+        solveTest2 @?= [dInt]
+    , testCase "3: D [Maybe Int] (complete D, partial C)" $
+        solveTest3 @?= [cInt]
+    , testCase "4: C [[[Bool]]] (completeRules)" $
+        solveTest4 @?= []
+    , testCase "5: C (Either (Either Int Bool) (Either Bool Int)) (completeRules)" $
+        solveTest5 @?= []
+    , testCase "6: C (Either l x) (completeRules)" $
+        solveTest6 @?= [cL, cX]
+    , testCase "7: Sum test (completeRules)" $
+        solveTest7 @?= []
+    , testCase "8: Sum test (partialRules)" $
+        solveTest8 @?= [cBool, cInt]
+    , testCase "9: Rec test (completeRules)" $
+        solveTest9 @?= []
+    , testCase "10: Rec test (partialRules)" $
+        solveTest10 @?= [cBool, cInt]
+    , testCase "11: Prod test (completeRules)" $
+        solveTest11 @?= []
+    , testCase "12: Prod test (partialRules)" $
+        solveTest12 @?= [cBool, cInt]
+    , testCase "13: Dec test (completeRules)" $
+        solveTest13 @?= []
+    , testCase "14: Dec test (partialRules)" $
+        solveTest14 @?= [cBool, cInt]
+    ]
+  where
+    cInt = C _c Int
+    cBool = C _c Bool
+    cL = C _c _l
+    cX = C _c _x
+    dInt = C _d Int
+
+-- hardcoded variables, easier to read than (VarP "blah") everywhere
+_x, _xs, _vars, _name, _l, _body :: Pat
+_x = VarP "x"
+_xs = VarP "xs"
+_vars = VarP "vars"
+_name = VarP "name"
+_l = VarP "l"
+_body = VarP "body"
+
+-- TODO(@gnumonik): Import this from somewhere else when that somewhere else module exists
+mkStructuralRules :: Class -> [Rule]
+mkStructuralRules c =
+  [ C c Nil :<= []
+  , C c (_x :* _xs) :<= [C c _x, C c _xs]
+  , C c (_l := _x) :<= [C c _x]
+  , C c (RecP _xs) :<= [C c _xs]
+  , C c (ProdP _xs) :<= [C c _xs]
+  , C c (SumP _xs) :<= [C c _xs]
+  , C c (DecP _name _vars _body) :<= [C c _body]
+  ]
```

Comment:
Imo our structural rules for types of kind `*` should be

```prolog
% canonical representation of TyBody
class(unit).
class(void).
class(either(A,B)) :- class(A), class(B).
class(prod(A,B)) :- class(A), class(B).

% type application
class(ty_app(F, A)) :- apply(F, A, Res), class(Res).

% ty_ref
class(ty_ref(Ref)) :- ty_def(Ref, Ty), class(Ty).

% and perhaps for 0 arg abstractions
class(ty_abs([], TyBody)) :- class(TyBody).
```
There are 2 special rules that deal with lambda application (unwrapping the type def) and fetching a ty ref from the context.

Reply:
Read on how that affects things.

Reply:
Completely agree that your structural rules are identical to canonical ones, I just used them from brevity.

Reply:
The rules for TyRef and TyApp are essential, and them being absent is a signal we're a bit on the hacky side of things.

A typeclass Eq is declaring that it can be applied to types of kind `*`. What are those types?
1. Sum/Prod/Nil
2. TyApp - of kind *
3. TyRef - of kind *
4. TyAbs - of kind *

The things you're missing and should be added explicitly:

```prolog
% type application
class(ty_app(F, A)) :- apply(F, A, Res), class(Res).

% ty_ref
class(ty_ref(Ref)) :- ty_def(Ref, Ty), class(Ty).

% and perhaps for 0 arg abstractions
class(ty_abs([], TyBody)) :- class(TyBody).
```

What would `apply` and `ty_def` be in your formulation?

```haskell
data Rule where
  (:<=) :: Constraint -> [Constraint] -> Rule

data Constraint = C Class Pat | Apply Pat Pat Pat | TyDef TyRef Ty
```

You just extended your language to contain `apply` that performs 'lambda application' and `ty def` that just queries the 'context' for the requested Ty term bound to this TyRef. (which should fail btw with an API error)

I imagine that `solve` would then (i guess) use 'subV' the implement this call.

Reply:
...I just remembered that `DecP` only exists because I was using patterns in the codegen and needed a pattern for the whole declaration. So yeah it's not really necessary, I guess. 

I suspect that there isn't something weird about what you're proposing too, though. Don't take the rest of this post as my final word on the matter b/c it might take me more than one try to put my finger on it. I'm just gonna think out loud: 

- `TyRef`s are just an identifier, right? If we were actually _evaluating_ anything here (we're not), we'd replace them with the body of the TyDef they refer to and then alpha conversion and beta reduction. They're morally equivalent to named functions in a lambda calculus system - i.e., they're necessary for user experience/sanity but from a certain point of view aren't necessary. 

- But they're also, kind of, not an identifier in that they're our version of `TyCon`. 

- I don't understand what `Res` is because we aren't actually evaluating. When you evaluate a function application in the lambda calculus you get a value (...me, just now realizing that's why it's called evaluation...). So `Res` would be the value that results from applying `F` to `a` for some (type) function `F` and some concrete type (`:: *`) `a`. I think what's confusing is me is that evaluating `Maybe Int` (i.e. `AppP (RefP _ (Name "Maybe")) (RefP _ (Name "Int))` gives us... `Maybe Int` (i.e. `AppP (RefP _ (Name "Maybe")) (RefP _ (Name "Int))`. 

- `A typeclass Eq is declaring that it can be applied to types of kind *.` That sounds right, but....

> 
    Sum/Prod/Nil
    TyApp - of kind *
    TyRef - of kind *
    TyAbs - of kind *
>

Is that really correct? I'm honestly not sure. 

I don't think `Sum/Prod/Nil` are really "types of kind *". It seems to me that they're the _values_ of _some concrete type_ which is of kind *.  If they're types then they definitely aren't of kind *, `Sum` has the kind `[[*]]` and `Prod` has the kind `[*]`.

Reply:
``` haskell
data Constraint = C Class Pat | Apply Pat Pat Pat | TyDef TyRef Ty
```

I have an overwhelmingly strong intuition that this isn't right. 

I can interpret `Apply` as a constraint if it's an equality constraint, e.g. `Apply F a ~ (F a)`. However the only thing we ever apply is a TyRef to a type w/ TyVar arguments, we never actually perform any reduction or normalization, and so the equality is trivial. I think in every possible case where we could use it, it's just `F a ~ F a` (or something that's a substitution instance of that).

Reply:
To put the last point in a slightly different way: `solve` is the purely logical component of a larger system which has procedural components. `solve` emits subgoals that could not be determined on the basis of the rules that it has access to when we feed it a particular constraint. That's all it does / all it's meant to do. 

The `checkDerive` function that used to be in `Utils` contains the extra-logical components (resolving references & checking whether they're opaque, construcing the scope to feed to `solve`, etc). 

What you're suggesting is that we embed *everything* into the pure logical context. And that makes a _lot_ of sense if you're nourishing yourself on a diet of prolog because everything (more or less) in prolog only exists in a purely logical context of relations. I just think that embedding the resolution of `TyRef`s into rules doesn't make sense in Haskell; using `M.lookup` in a function that orchestrates invocations of `solve` strikes me as perfectly fine.